src/stats.c

```diff
@@ -0,0 +1,364 @@
+#include "stats.h"
+
+#define min(a, b) ((a) < (b) ? (a) : (b))
+
+
+/**
+ * @brief Calculates the maximum element of an array
+ * @param array array containing the values needed to calculate the maximum
+ * @param size  size of the array
+ */
+real_number maxOfArray(real_number* array, vec_size size) {
+
+  real_number max = MIN_REAL_NUMBER;
+  for (vec_size i = 0; i < size; ++i) {
+    if (array[i] > max) {
+      max = array[i];
+    }
+  }
+  return max;
+}
+
+/**
+ * @brief Calculates the minimum element of an array
+ * @param array array containing the values needed to calculate the minimum
+ * @param size  size of the array
+ */
+real_number minOfArray(real_number* array, vec_size size) {
+
+  real_number min = MAX_REAL_NUMBER;
+  for (vec_size i = 0; i < size; ++i) {
+    if (array[i] < min) {
+      min = array[i];
+    }
+  }
+  return min;
+}
+
+/**
+ * @brief Calculates the mean of a given array
+ * @param data  array containing the values needed to calculate the mean
+ * @param size  size of the array
+ */
+real_number mean(real_number* data, vec_size size) {
+
+  real_number sum = 0.0;
+  for (vec_size i = 0; i < size; ++i) {
+    sum += data[i];
+  }
+  return sum / size;
+}
+
+/**
+ * @brief Computes the variance of a sample
+ * (divides by [size - 1] since the given values are considered to be a sample
+ * of the population)
+ * @param data  array containing the values of the sample
+ * @param size  number of samples
+ * @return      variance of sample
+ */
+real_number variance(real_number* data, vec_size size) {
+
+  real_number avg = mean(data, size);
+  real_number var = 0.0;
+  real_number adujstingFactor = 0.0;
+  for (vec_size i = 0; i < size; ++i) {
+    real_number diffElem = data[i] - avg;
+    adujstingFactor += diffElem;
+    var += diffElem * diffElem;
+  }
+
+  return (var - adujstingFactor * adujstingFactor / size) / (size - 1);
+}
+
+/**
+ * @brief Computes the standard deviation of a sample
+ * (divides by [size - 1] since the given values are considered to be a sample
+ * of the population)
+ * @param array array containing the values of the sample
+ * @param size  number of samples
+ * @return standard deviation of sample
+ */
+real_number standardDeviation(real_number* array, vec_size size) {
+  return sqrt(variance(array, size));
+}
+
+/**
+ * @brief Computes the absolute average deviation of a sample
+ * (divides by [size - 1] since the given values are considered to be a sample
+ * of the population)
+ * @param array array containing the values of the sample
+ * @param size  number of samples
+ * @return absolute average deviation of sample
+ */
+real_number absAverageDeviation(real_number* array, vec_size size) {
+  real_number avg = mean(array, size);
+  real_number result = 0.0;
+  for (vec_size i = 0; i < size; ++i) {
+    real_number elem = array[i];
+    real_number diffElem = elem - avg;
```

Comment:
Is the elem variable necessary. Cant you directly use array[i] to calculate diffElem